### PR TITLE
homed: add ability to auto-grow home dir in login and auto-shrink on logout (if luks2+btrfs is used)

### PR DESCRIPTION
**Pull request by poettering**
Also split out of #21253 

(Note that this is enabled by default, hence the existing tests will automatically start testing this)

### CODE REVIEW

File: src/home/homework-luks.c [hunk 1]
```diff
@@ -247,15 +248,59 @@ static int run_fsck(const char *node, const char *fstype) {
...
+         * around, since we'll need it when automatically resizing (since we can't ask the user there
+         * again). We do this by uploading it into the kernel keyring, specifically the "session" one. This
+         * is done under the assumption systemd-homed gets its private per-session keyring (i.e. default
+         * service behaviour, given that KeyringMode=private is the default). It will survive between our
+         * systemd-homework invocations that way.
+         *
+         * If auto-shrink-on-logout is disabled we'll skip this step, to be frugal with sensitive data. */
+
+        if (user_record_auto_resize_mode(h) != AUTO_RESIZE_SHRINK_AND_GROW) {  /* Won't need it */
+                *ret_key_serial = -1;
```
Comment (bluca) [inline src/home/homework-luks.c]: `if (ret_key_serial)...`

File: src/home/homework-luks.c [hunk 2]
```diff
@@ -3540,6 +3608,13 @@ int home_passwd_luks(
                         return log_error_errno(r, "Failed to set up LUKS password: %m");
 
                 log_info("Updated LUKS key slot %zu.", i);
+
+                /* If we changed the password, then make sure to update the copy in the keyring, so that
+                 * auto-rebalance continues to work. We only do this if we operate on an active home dir. */
+                if (i == 0 && FLAGS_SET(flags, HOME_SETUP_ALREADY_ACTIVATED)) {
+                        key_serial_t serial = -1;
+                        upload_to_keyring(h, effective_passwords[i], &serial);
```
Comment (bluca) [inline src/home/homework-luks.c]: `serial` is not used? Why not simply omit it?

**Review by bluca — APPROVED** [review]
lgtm, just two minor things

**Comment by poettering** [comment]
force pushed new version with the two issues fixed. no other changes. upgrading green label.

**Comment by bluca** [comment]
```
08:59:31 FAILED TASKS:
08:59:31 -------------
08:59:31 TEST-56-EXIT-TYPE
```

This is new I think, but I can't see how it's related - @mrc0mmand have you seen that fail before?

**Comment by mrc0mmand** [comment]
> ```
> 08:59:31 FAILED TASKS:
> 08:59:31 -------------
> 08:59:31 TEST-56-EXIT-TYPE
> ```
> 
> This is new I think, but I can't see how it's related - @mrc0mmand have you seen that fail before?

I've seen it a couple of times since the test was reintroduced, but so far didn't find time to investigate it further. So it's definitely not related to this PR.

**Comment by poettering** [comment]
Hmmm, the focal-amd64 CI timed out, which I guess is unrelated, but let's merge since the homed test case ran there on the previous push and passed.

**Pull request merged**